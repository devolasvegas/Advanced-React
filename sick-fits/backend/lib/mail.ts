import { createTransport, getTestMessageUrl } from 'nodemailer';

const transport = createTransport({
  host: process.env.MAIL_HOST,
  port: process.env.MAIL_PORT,
  auth: {
    user: process.env.MAIL_USER,
    pass: process.env.MAIL_PASS,
  },
});

function makeANiceEmail(text: string): string {
  return `
    <div style="border: 1px solid black; padding: 20px; font-family: sans-serif; line-height: 2; font-size: 20px;">
      <h2>Hello There!</h2>
      <p>${text}</p>
      <p>ðŸ˜˜, The Team</p>
    </div>
  `;
}

export async function sendPasswordResetEmail(
  resettoken: string,
  to: string
): Promise<void> {
  // send user reset token email
  const info = await transport.sendMail({
    to,
    from: 'noreply@devondaviau.com',
    subject: 'Your Password Reset Token',
    html: makeANiceEmail(
      `Your Password Reset Token is here! \n\n <a href="${process.env.FRONTEND_URL}/reset?token=${resettoken}">Click Here to Reset</a>`
    ),
  });

  if (process.env.MAIL_USER.includes('ethereal.email')) {
    console.log('Password reset email sent successfully to:', to);
    console.log(
      `ðŸ’Œ Message Sent! Preview it at: ${getTestMessageUrl(info) as string}`
    );
  }
}
