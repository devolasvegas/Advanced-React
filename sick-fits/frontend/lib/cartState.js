import { createContext, useContext, useState } from 'react';
import PropTypes from 'prop-types';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // This is our own custom state provider. We will store data (state) and functionality (setters and getters) here and anyone can access it via the consumer!
  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen((prev) => !prev);
  }
  function closeCart() {
    setCartOpen(false);
  }
  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };

CartStateProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
