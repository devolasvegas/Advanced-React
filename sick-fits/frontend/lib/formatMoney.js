export default function formatMoney(cents = 0) {
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  };

  // If amount is a whole number of dollars, remove decimal places
  if (cents % 100 === 0) {
    options.minimumFractionDigits = 0;
    options.maximumFractionDigits = 0;
  }

  const formatter = new Intl.NumberFormat('en-US', options);
  return formatter.format(cents / 100);
}
