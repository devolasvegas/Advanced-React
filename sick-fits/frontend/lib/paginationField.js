import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // Tell Apollow we will be handling caching manually
    read(existing = [], { args, cache }) {
      const { skip, first } = args;

      // Read the number of items on the page from the cache
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count || 0;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // Check if we have existing items
      const items = existing.slice(skip, skip + first).filter((x) => x);

      // If
      // There are items
      // AND there arent enough items to satisfy how many were requestes
      // AND we are on the last page
      // THEN JUST SEND IT
      // if (items.length && items.length !== first && page === pages) {
      //   return items;
      // }

      // if (items.length !== first) {
      //   return false; // We don't have any items, so we need to go to the network to fetch them
      // }

      // If there are items, we return the items we have, and we don't have to go to the network
      if (items.length) {
        return items;
      }

      // Fallback to network
      return false;
    },
    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      const merged = existing ? existing.slice(0) : [];

      for (let i = skip; i < skip + incoming.length; i += 1) {
        merged[i] = incoming[i - skip];
      }

      // return merged array
      return merged;
    },
  };
}
