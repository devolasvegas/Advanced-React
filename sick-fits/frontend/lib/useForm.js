import { useState } from 'react';

export default function useForm(initial = {}) {
  const [inputs, setInputs] = useState(initial);

  function handleChange(e) {
    let { value, type, name } = e.target;

    // Handle number inputs
    // Parse string value from input to a number for insertion into the DB
    if (type === 'number') {
      value = parseFloat(value);
    }

    // Handle file inputs
    if (type === 'file') {
      [value] = e.target.files;
    }
    // Spread in existing state, and add our new input name and value
    setInputs({
      ...inputs,
      [name]: value,
    });
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const blankState = Object.fromEntries(
      Object.keys(inputs).map((key) => [key, ''])
    );

    setInputs(blankState);
  }

  // Return the things we want to surface from this custom hook
  return {
    inputs,
    handleChange,
    resetForm,
    clearForm,
  };
}
