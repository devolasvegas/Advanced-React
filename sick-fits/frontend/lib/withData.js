import { ApolloClient, ApolloLink, InMemoryCache } from '@apollo/client';
import { onError } from '@apollo/link-error';
import { getDataFromTree } from '@apollo/client/react/ssr';
import { createUploadLink } from 'apollo-upload-client';
import withApollo from 'next-with-apollo';
import paginationField from './paginationField';

function createClient({ headers, initialState }) {
  const isServer = typeof window === 'undefined';

  const endPoint = isServer
    ? process.env.PRIVATE_BACKEND_URL // private network URL
    : process.env.NEXT_PUBLIC_BACKEND_URL; // public frontend-accessible URL

  if (!endPoint) {
    console.error('Missing endpoint URL!');
  }

  return new ApolloClient({
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors)
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          );
        if (networkError)
          console.log(
            `[Network error]: ${networkError}. Backend is unreachable. Is it running?`
          );
      }),
      // this uses apollo-link-http under the hood, so all the options here come from that package
      createUploadLink({
        uri: endPoint,
        fetchOptions: {
          credentials: 'include',
        },
        // pass the headers along from this request. This enables SSR with logged in state
        headers,
      }),
    ]),
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            // Handle pagination for products manually
            allProducts: paginationField(),
          },
        },
      },
    }).restore(initialState || {}),
  });
}

export default withApollo(createClient, { getDataFromTree });
