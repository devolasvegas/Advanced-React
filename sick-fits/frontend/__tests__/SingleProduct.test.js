import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';

import SingleProduct, {
  SINGLE_PRODUCT_QUERY,
} from '../components/SingleProduct';

import { fakeItem } from '../lib/testUtils';

const product = fakeItem();

const mocks = [
  {
    request: {
      query: SINGLE_PRODUCT_QUERY,
      variables: { id: 'abc123' },
    },
    result: {
      data: {
        Product: product,
      },
    },
  },
];

describe('<SingleProduct />', () => {
  it('renders with proper data', async () => {
    // We need some fake data
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SingleProduct id="abc123" />
      </MockedProvider>
    );

    await screen.findByTestId('singleProduct');
    expect(container).toMatchSnapshot();
    // screen.debug();
    // screen.logTestingPlaygroundURL();
  });

  it('Displays error when an item is not found', async () => {
    const errorMocks = [
      {
        request: {
          query: SINGLE_PRODUCT_QUERY,
          variables: { id: 'abc123' },
        },
        result: {
          errors: [{ message: 'Item not found!' }],
        },
      },
    ];

    const { container, debug } = render(
      <MockedProvider mocks={errorMocks}>
        <SingleProduct id="abc123" />
      </MockedProvider>
    );

    await screen.findByTestId('graphql-error');

    expect(container).toHaveTextContent('Shoot!');
    expect(container).toHaveTextContent('Item not found!');
  });
});
