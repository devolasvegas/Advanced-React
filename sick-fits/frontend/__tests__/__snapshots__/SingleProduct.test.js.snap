// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleProduct /> renders with proper data 1`] = `
<div>
  <div
    class="SingleProduct__ProductStyles-sc-1kapz8w-0 huslPg"
    data-testid="singleProduct"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      class="details"
    >
      <h2>
        dogs are best
      </h2>
      <p>
        dogs
      </p>
      <p>
        5000
      </p>
    </div>
  </div>
</div>
`;
